File: modules/cyc-bridge/ios/CycProtocol.swift
import Foundation

enum CycError: LocalizedError {
  case invalid(String)
  var errorDescription: String? { if case .invalid(let message) = self { return message }; return nil }
}

enum CycRequest: UInt8 {
  case identity = 111
  case selective = 50

  var frame: Data {
    let payload: [UInt8] = self == .identity ? [rawValue] : [rawValue] + withUnsafeBytes(of: CycProtocol.selectedMask.bigEndian, Array.init)
    let crc = CycProtocol.crc(payload)
    return Data([2, UInt8(payload.count)] + payload + [UInt8(crc >> 8), UInt8(crc & 255), 3])
  }
}

// Both families use the same published selective layout, verified against live X6/X12 replies.
// Keep the adapter allowlist explicit; a similar model name is not proof of compatibility.
enum CycControllerAdapter: String, Codable {
  case x6 = "X6", x12 = "X12"

  func decodeTelemetry(_ payload: [UInt8]) throws -> [String: Double] {
    switch self {
    case .x6, .x12: return try CycProtocol.decodeTelemetry(payload)
    }
  }
}

struct CycControllerIdentity: Codable {
  let controllerModel: String
  let firmwareLabel: String
  let major: UInt8
  let minor: UInt8
  let adapter: CycControllerAdapter

  var hasKnownSpeedUnit: Bool { (controllerModel == "X6" || controllerModel == "X12") && major == 5 && minor == 3 }

  func annotate(_ sample: [String: Any]) -> [String: Any] {
    var result = sample
    result["controllerModel"] = controllerModel
    result["firmwareLabel"] = firmwareLabel
    result["controllerProtocol"] = "\(major).\(minor)"
    // CYC 5.3 field 23 sends app_get_speed() in km/h, regardless of display units.
    result["controllerSpeedMps"] = hasKnownSpeedUnit ? (sample["speedRaw"] as? Double).map { $0 / 3.6 } : nil
    return result
  }
}

enum CycProtocol {
  static let selectedMask: UInt32 = 0x03c0fb8f
  static let maximumGap = 2.5
  static let columns = ["timestamp", "elapsedSeconds", "sequence", "humanPowerW", "cadenceRpm",
    "motorInputPowerW", "batteryVoltageV", "batteryCurrentA", "motorCurrentA", "motorRpm",
    "pedalTorqueNm", "controllerTempC", "motorTempC", "consumedAh", "consumedWh",
    "throttleVoltageV", "faultCode", "assistLevel", "raceMode", "speedRaw",
    "controllerSpeedMps", "controllerModel", "firmwareLabel", "controllerProtocol"]
  static let csvHeader = columns.joined(separator: ",")

  static func crc(_ bytes: [UInt8]) -> UInt16 {
    var value: UInt16 = 0
    for byte in bytes {
      value ^= UInt16(byte) << 8
      for _ in 0..<8 { value = (value & 0x8000) != 0 ? (value &<< 1) ^ 0x1021 : value &<< 1 }
    }
    return value
  }

  // Only the ASCII model/firmware prefix is a name. Real CYC responses can contain
  // opaque controller bytes before the terminating NUL; never decode/expose that tail.
  @discardableResult
  static func validateIdentity(_ payload: [UInt8]) throws -> CycControllerIdentity {
    let unsupported = CycError.invalid("Unsupported controller. Connect a CYC X6 or X12.")
    guard (4...1024).contains(payload.count), payload[0] == 111 || payload[0] == 0,
      let end = payload[3...].firstIndex(of: 0),
      end - 3 <= 128 else { throw unsupported }
    let ascii = payload[3..<end].prefix { (0x20...0x7e).contains($0) }
    guard let text = String(bytes: ascii, encoding: .ascii),
      let prefix = text.range(of: "^X(6|12)([A-Za-z_][A-Za-z0-9_]{0,29})? +[0-9]{6,8}[A-Z]{0,8}", options: .regularExpression) else { throw unsupported }
    let boundary = 3 + text.distance(from: text.startIndex, to: prefix.upperBound)
    // A partial date/model match must not turn unsupported bytes into a valid prefix.
    guard boundary == end || payload[boundary] == 0x20 else { throw unsupported }
    let parts = text[prefix].split(separator: " ")
    let model = String(parts[0])
    return CycControllerIdentity(controllerModel: model, firmwareLabel: String(parts[1]),
      major: payload[1], minor: payload[2], adapter: model.hasPrefix("X12") ? .x12 : .x6)
  }

  static func decodeTelemetry(_ payload: [UInt8]) throws -> [String: Double] {
    // Exact selected layout: 5-byte command/mask + 49 data bytes.
    guard payload.count == 54, payload.first == CycRequest.selective.rawValue else {
      throw CycError.invalid("Telemetry command or length does not match the verified profile.")
    }
    var cursor = 1
    func unsigned(_ width: Int) -> UInt32 {
      defer { cursor += width }
      return payload[cursor..<(cursor + width)].reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
    }
    guard unsigned(4) == selectedMask else { throw CycError.invalid("Unexpected telemetry field mask.") }
    func signed(_ width: Int, _ divisor: Double = 1) -> Double {
      let raw = unsigned(width)
      return (width == 2 ? Double(Int16(bitPattern: UInt16(raw))) : Double(Int32(bitPattern: raw))) / divisor
    }
    var v: [String: Double] = [:]
    v["controllerTempC"] = signed(2, 10)
    v["motorTempC"] = signed(2, 10)
    v["motorCurrentA"] = signed(4, 100)
    v["batteryCurrentA"] = signed(4, 100)
    v["motorRpm"] = signed(4)
    v["batteryVoltageV"] = signed(2, 10)
    v["consumedAh"] = signed(4, 10_000)
    v["consumedWh"] = signed(4, 10_000)
    v["cadenceRpm"] = signed(4, 10_000)
    v["throttleVoltageV"] = signed(4, 100)
    v["pedalTorqueNm"] = signed(4, 100)
    v["faultCode"] = Double(unsigned(1))
    v["humanPowerW"] = signed(4)
    v["speedRaw"] = signed(4, 100)
    v["raceMode"] = Double(unsigned(1))
    v["assistLevel"] = Double(unsigned(1))
    guard cursor == payload.count else { throw CycError.invalid("Unknown trailing telemetry data.") }
    v["motorInputPowerW"] = v["batteryVoltageV"]! * v["batteryCurrentA"]!
    return v
  }

  private static let timestampLock = NSLock()
  private static let timestampFormatter: ISO8601DateFormatter = {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    formatter.timeZone = TimeZone(secondsFromGMT: 0)
    return formatter
  }()
  static func timestamp(_ date: Date = Date()) -> String {
    timestampLock.lock(); defer { timestampLock.unlock() }
    return timestampFormatter.string(from: date)
  }
}

struct CycFrameDecoder {
  private(set) var buffer: [UInt8] = []
  private(set) var discardedBytes = 0
  mutating func reset() { discardedBytes += buffer.count; buffer.removeAll(keepingCapacity: true) }

  mutating func feed(_ data: Data) -> [[UInt8]] {
    buffer.append(contentsOf: data)
    var packets: [[UInt8]] = []
    while !buffer.isEmpty {
      var incomplete: Int?
      var found = false
      for offset in buffer.indices {
        let start = buffer[offset]
        guard start == 2 || start == 3 else { continue }
        let header = start == 2 ? 2 : 3
        let remaining = buffer.count - offset
        guard remaining >= header else { if incomplete == nil { incomplete = offset }; continue }
        let length = buffer[(offset + 1)..<(offset + header)].reduce(0) { ($0 << 8) | Int($1) }
        guard (1...1024).contains(length), start != 3 || length > 255 else { continue }
        let size = header + length + 3
        guard remaining >= size else { if incomplete == nil { incomplete = offset }; continue }
        let payload = Array(buffer[(offset + header)..<(offset + header + length)])
        let crc = UInt16(buffer[offset + size - 3]) << 8 | UInt16(buffer[offset + size - 2])
        guard buffer[offset + size - 1] == 3, CycProtocol.crc(payload) == crc else { continue }
        packets.append(payload)
        discardedBytes += offset
        buffer.removeFirst(offset + size)
        found = true
        break
      }
      if found { continue }
      let discarded = incomplete ?? buffer.count
      discardedBytes += discarded
      buffer.removeFirst(discarded)
      break
    }
    // The only retained data is an incomplete frame (at most 1028 bytes).
    if buffer.count > 1028 { reset() }
    return packets
  }
}

File: modules/cyc-bridge/android/src/main/java/app/powerlog/bridge/CycProtocol.kt
package app.powerlog.bridge

import java.nio.ByteBuffer
import java.nio.ByteOrder

/** The only outbound controller operations are the two read requests below. */
object CycProtocol {
    const val SERVICE = "6e400001-b5a3-f393-e0a9-e50e24dcca9e"
    const val WRITE = "6e400002-b5a3-f393-e0a9-e50e24dcca9e"
    const val NOTIFY = "6e400003-b5a3-f393-e0a9-e50e24dcca9e"
    const val MASK = 0x03c0fb8f

    enum class Read {
        IDENTITY,
        TELEMETRY,
    }

    fun crc(bytes: ByteArray): Int {
        var crc = 0
        for (byte in bytes) {
            crc = crc xor ((byte.toInt() and 255) shl 8)
            repeat(8) { crc = ((crc shl 1) xor if (crc and 0x8000 != 0) 0x1021 else 0) and 65535 }
        }
        return crc
    }

    fun request(read: Read): ByteArray {
        val payload =
            if (read == Read.IDENTITY) byteArrayOf(111)
            else byteArrayOf(50, 3, 0xc0.toByte(), 0xfb.toByte(), 0x8f.toByte())
        val crc = crc(payload)
        return byteArrayOf(2, payload.size.toByte()) +
            payload +
            byteArrayOf((crc shr 8).toByte(), crc.toByte(), 3)
    }

    data class Identity(val model: String, val firmware: String, val protocol: String) {
        val knownSpeed
            get() = model in listOf("X6", "X12") && protocol == "5.3"
    }

    fun identity(bytes: ByteArray): Identity {
        require(bytes.size in 5..1024 && (bytes[0].toInt() and 255) in listOf(0, 111)) {
            "Unsupported controller"
        }
        val end =
            (3 until bytes.size).firstOrNull { bytes[it] == 0.toByte() }
                ?: error("Invalid controller identity")
        require(end - 3 in 1..128)
        val text = bytes.copyOfRange(3, end).toString(Charsets.US_ASCII)
        val match =
            Regex("^(X(?:6|12)(?:[A-Za-z_][A-Za-z0-9_]{0,29})?) +([0-9]{6,8}[A-Z]{0,8})(?= |$)")
                .find(text) ?: error("Unsupported controller. Connect a CYC X6 or X12.")
        return Identity(
            match.groupValues[1],
            match.groupValues[2],
            "${bytes[1].toInt() and 255}.${bytes[2].toInt() and 255}",
        )
    }

    private data class Field(val name: String, val bytes: Int, val scale: Double)

    private val fields =
        listOf(
            Field("controllerTempC", 2, 10.0),
            Field("motorTempC", 2, 10.0),
            Field("motorCurrentA", 4, 100.0),
            Field("batteryCurrentA", 4, 100.0),
            Field("idCurrentA", 4, 100.0),
            Field("iqCurrentA", 4, 100.0),
            Field("dutyCycle", 2, 1000.0),
            Field("motorRpm", 4, 1.0),
            Field("batteryVoltageV", 2, 10.0),
            Field("consumedAh", 4, 10000.0),
            Field("tripTimeRaw", 4, 1.0),
            Field("consumedWh", 4, 10000.0),
            Field("cadenceRpm", 4, 10000.0),
            Field("throttleVoltageV", 4, 100.0),
            Field("pedalTorqueNm", 4, 100.0),
            Field("faultCode", 1, 1.0),
            Field("ioFlags", 4, 1.0),
            Field("controllerId", 1, 1.0),
            Field("temperatures", 6, 10.0),
            Field("vdV", 4, 1000.0),
            Field("vqV", 4, 1000.0),
            Field("odometerRaw", 4, 1.0),
            Field("humanPowerW", 4, 1.0),
            Field("speedRaw", 4, 100.0),
            Field("raceMode", 1, 1.0),
            Field("assistLevel", 1, 1.0),
        )

    fun telemetry(bytes: ByteArray, identity: Identity): Map<String, Double> {
        val buffer = ByteBuffer.wrap(bytes).order(ByteOrder.BIG_ENDIAN)
        require(buffer.remaining() >= 5 && buffer.get() == 50.toByte() && buffer.int == MASK) {
            "Unexpected telemetry response"
        }
        val values = linkedMapOf<String, Double>()
        fields.forEachIndexed { bit, field ->
            if (MASK and (1 shl bit) != 0) {
                require(buffer.remaining() >= field.bytes) { "Truncated telemetry" }
                val raw =
                    when (field.bytes) {
                        1 -> buffer.get().toInt() and 255
                        2 -> buffer.short.toInt()
                        else -> buffer.int
                    }
                values[field.name] = raw / field.scale
            }
        }
        require(!buffer.hasRemaining()) { "Unexpected telemetry length" }
        values["motorInputPowerW"] =
            Math.round(
                values.getValue("batteryVoltageV") * values.getValue("batteryCurrentA") * 10000
            ) / 10000.0
        if (identity.knownSpeed) values["controllerSpeedMps"] = values.getValue("speedRaw") / 3.6
        return values
    }

    class Decoder {
        private var buffer = byteArrayOf()
        var discarded = 0L
            private set

        fun reset() {
            discarded += buffer.size
            buffer = byteArrayOf()
        }

        fun feed(bytes: ByteArray): List<ByteArray> {
            require(bytes.size <= 65536)
            buffer += bytes
            val result = mutableListOf<ByteArray>()
            while (buffer.isNotEmpty()) {
                var incomplete: Int? = null
                var found = false
                for (offset in buffer.indices) {
                    val start = buffer[offset].toInt() and 255
                    if (start != 2 && start != 3) continue
                    val header = if (start == 2) 2 else 3
                    if (buffer.size - offset < header) {
                        if (incomplete == null) incomplete = offset
                        continue
                    }
                    val length =
                        if (start == 2) buffer[offset + 1].toInt() and 255
                        else
                            ((buffer[offset + 1].toInt() and 255) shl 8) or
                                (buffer[offset + 2].toInt() and 255)
                    if (length !in 1..1024 || start == 3 && length <= 255) continue
                    val size = header + length + 3
                    if (buffer.size - offset < size) {
                        if (incomplete == null) incomplete = offset
                        continue
                    }
                    val payload = buffer.copyOfRange(offset + header, offset + header + length)
                    val pos = offset + header + length
                    val checksum =
                        ((buffer[pos].toInt() and 255) shl 8) or (buffer[pos + 1].toInt() and 255)
                    if (buffer[offset + size - 1] != 3.toByte() || crc(payload) != checksum)
                        continue
                    result.add(payload)
                    discarded += offset
                    buffer = buffer.copyOfRange(offset + size, buffer.size)
                    found = true
                    break
                }
                if (found) continue
                val drop = incomplete ?: buffer.size
                discarded += drop
                buffer = buffer.copyOfRange(drop, buffer.size)
                break
            }
            return result
        }
    }
}
